hail/src/main/scala/is/hail/expr/types/physical/PCanonicalBaseStruct.scala

```diff
@@ -0,0 +1,223 @@
+  }
+
+  def deepPointerCopy(region: Region, dstStructAddress: Long) {
+    var i = 0
+    while(i < this.size) {
+      val dstFieldType = this.fields(i).typ.fundamentalType
+      if(dstFieldType.containsPointers && this.isFieldDefined(dstStructAddress, i)) {
+        val dstFieldAddress = this.fieldOffset(dstStructAddress, i)
+        dstFieldType match {
+          case t@(_: PBinary | _: PArray) =>
+            Region.storeAddress(dstFieldAddress, t.copyFromType(region, dstFieldType, Region.loadAddress(dstFieldAddress), forceDeep = true))
+          case t: PCanonicalBaseStruct =>
+            t.deepPointerCopy(region, dstFieldAddress)
+        }
+      }
+      i += 1
+    }
+  }
+
+  def copyFromType(mb: MethodBuilder, region: Code[Region], srcPType: PType, srcStructAddress: Code[Long], forceDeep: Boolean): Code[Long] = {
+    val sourceType = srcPType.asInstanceOf[PBaseStruct]
+    assert(sourceType.size == this.size)
+
+    if (this == sourceType && !forceDeep)
+      srcStructAddress
+    else {
+      val addr = mb.newLocal[Long]
+      Code(
+        addr := allocate(region),
+        constructAtAddress(mb, addr, region, sourceType, srcStructAddress, forceDeep),
+        addr
+      )
+    }
+  }
+
+  def copyFromTypeAndStackValue(mb: MethodBuilder, region: Code[Region], srcPType: PType, stackValue: Code[_], forceDeep: Boolean): Code[_] =
+    this.copyFromType(mb, region, srcPType, stackValue.asInstanceOf[Code[Long]], forceDeep)
+
+  def copyFromType(region: Region, srcPType: PType, srcStructAddress: Long, forceDeep: Boolean): Long = {
+    val sourceType = srcPType.asInstanceOf[PBaseStruct]
+    if (this == sourceType && !forceDeep)
+      srcStructAddress
+    else {
+      val newAddr = allocate(region)
+      constructAtAddress(newAddr, region, sourceType, srcStructAddress, forceDeep)
+      newAddr
+    }
+  }
+
+  def constructAtAddress(mb: MethodBuilder, addr: Code[Long], region: Code[Region], srcPType: PType, srcAddress: Code[Long], forceDeep: Boolean): Code[Unit] = {
+    val srcStruct = srcPType.asInstanceOf[PBaseStruct]
+    val addrVar = mb.newLocal[Long]
+
+    if (srcStruct == this) {
+      var c: Code[Unit] = Code(
+        addrVar := addr,
+        Region.copyFrom(srcAddress, addrVar, byteSize))
+      if (forceDeep) {
+        c = Code(c, deepPointerCopy(mb, region, addrVar))
+      }
+      c
+    } else {
+      val srcAddrVar = mb.newLocal[Long]
+      Code(
+        srcAddrVar := srcAddress,
+        addrVar := addr,
+        stagedInitialize(addrVar, setMissing = true),
+        Code(fields.zip(srcStruct.fields).map { case (dest, src) =>
+          assert(dest.typ.required <= src.typ.required)
+          val idx = dest.index
+          assert(idx == src.index)
+          srcStruct.isFieldDefined(srcAddrVar, idx).orEmpty(Code(
+            setFieldPresent(addrVar, idx),
+            dest.typ.constructAtAddress(mb, fieldOffset(addrVar, idx), region, src.typ, srcStruct.loadField(srcAddrVar, idx), forceDeep))
```

Comment:
similarly, wondering why not called on the fundamentalType

Reply:
oops, made this comment yesterday and changed my mind. no reason this needs to be fundamental.